core/shared/src/main/scala/fs2/concurrent/Topic.scala

```diff
@@ -124,9 +125,21 @@ object Topic {
                 } <* subscriberCount.update(_ + 1)
 
                 def unsubscribe(id: Long) =
-                  state.update { case (subs, nextId) =>
-                    (subs - id, nextId)
-                  } >> subscriberCount.update(_ - 1)
+                  state.modify { case (subs, nextId) =>
+                    // _After_ we remove the bounded queue for this
+                    // subscriber, we need to drain it to unblock to
+                    // publish loop which might have already enqueued
+                    // something.
+                    def drainQueue: F[Unit] =
+                      subs.get(id).traverse_ { q =>
+                        q.tryTake.flatMap {
+                          case None    => F.unit
+                          case Some(_) => drainQueue
+                        }
```

Comment:
Perhaps another traverse? 
```suggestion
                        q.tryTake.flatMap(_.traverse_(_ => drainQueue))
```

Reply:
Used `iterateUntil`, which avoid redoing the `get` as well